lib/kernel.js

```diff
@@ -90,12 +90,32 @@ export default class Kernel {
     throw new Error("Kernel: restart method not implemented");
   }
 
-  execute(code: string, onResults: Function) {
-    throw new Error("Kernel: execute method not implemented");
+  _execute(code: string, callWatches: boolean, onResults: ?Function) {
+    throw new Error("Kernel: _execute method not implemented");
+  }
+
+  executeMaybeWatch(code: string, callWatches: boolean, onResults: ?Function) {
+    let cancelled = false;
+    this.emitter.emit("will-execute", {
+      code: code,
+      callWatches: callWatches,
+      onResults: onResults,
+      cancel: () => {
+        cancelled = true;
+      }
+    });
+
+    if (!cancelled) {
```

Comment:
Note that this code is incorrect if there are multiple listeners: as soon as one cancels the execution no more listeners down the chain should be run. The Emitter class doesn't provide this kind of event propagation behavior, so I'll need to figure out some alternative.